pkg/service/proofs/caching.go

```diff
@@ -0,0 +1,182 @@
+package proofs
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/url"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/storacha/go-libstoracha/capabilities/access"
+	"github.com/storacha/go-ucanto/client"
+	"github.com/storacha/go-ucanto/core/delegation"
+	"github.com/storacha/go-ucanto/core/invocation"
+	"github.com/storacha/go-ucanto/core/result"
+	"github.com/storacha/go-ucanto/did"
+	ucan_http "github.com/storacha/go-ucanto/transport/http"
+	"github.com/storacha/go-ucanto/ucan"
+)
+
+// defaultMinTTL is the minimum time a cached delegation should still be valid
+// for before it expires. If the TTL is less than this then consider it expired.
+var defaultMinTTL = time.Second * 5
+
+type CachingProofService struct {
+	cache      map[did.DID]map[did.DID]map[ucan.Ability]delegation.Delegation
```

Comment:
that's a lot of layers would be helpful to have a comment explaining what each key is

Reply:
looks like maybe issuer, then audience, then ability?

Reply:
Yes correct - will add comment.